File: FraudGraph-Agent/backend/app/schemas/case.py
from __future__ import annotations

from datetime import datetime
from enum import StrEnum
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from backend.app.schemas.investigation import RiskLevel


class CaseStatus(StrEnum):
    OPEN = "open"
    INVESTIGATING = "investigating"
    AWAITING_EVIDENCE = "awaiting_evidence"
    AWAITING_APPROVAL = "awaiting_approval"
    ACTIONED = "actioned"
    ESCALATED = "escalated"
    RESOLVED = "resolved"
    CLOSED = "closed"


class CaseOutcome(StrEnum):
    CONFIRMED_FRAUD = "confirmed_fraud"
    CLEARED = "cleared"
    INCONCLUSIVE = "inconclusive"
    ESCALATED = "escalated"


class DecisionType(StrEnum):
    ALLOW = "allow"
    BLOCK = "block"
    MONITOR = "monitor"
    WARN = "warn"
    REQUEST_EVIDENCE = "request_evidence"
    ESCALATE = "escalate"
    CREATE_REPORT = "create_report"
    CLOSE = "close"


class CaseFinding(BaseModel):
    model_config = ConfigDict(extra="forbid")

    finding_id: str

    title: str = Field(
        min_length=1,
    )

    description: str = Field(
        min_length=1,
    )

    evidence_ids: list[str] = Field(
        default_factory=list,
    )

    confidence: float | None = Field(
        default=None,
        ge=0.0,
        le=1.0,
    )

    created_at: datetime


class CaseDecision(BaseModel):
    model_config = ConfigDict(extra="forbid")

    decision_id: str

    decision_type: DecisionType

    rationale: str = Field(
        min_length=1,
    )

    evidence_ids: list[str] = Field(
        default_factory=list,
    )

    requires_approval: bool = False

    approved: bool | None = None

    approved_by: str | None = None

    created_at: datetime


class CaseAction(BaseModel):
    model_config = ConfigDict(extra="forbid")

    action_id: str

    action_type: str = Field(
        min_length=1,
    )

    status: str = Field(
        default="proposed",
        min_length=1,
    )

    rationale: str | None = None

    requires_approval: bool = False

    approval_id: str | None = None

    result: dict[str, Any] = Field(
        default_factory=dict,
    )

    created_at: datetime

    completed_at: datetime | None = None


class CaseMemoryReference(BaseModel):
    model_config = ConfigDict(extra="forbid")

    memory_id: str

    case_id: str

    similarity: float | None = Field(
        default=None,
        ge=0.0,
        le=1.0,
    )

    relevance_reason: str | None = None


class CaseCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    investigation_id: str

    transaction_id: str | None = None

    customer_id: str | None = None

    account_id: str | None = None

    title: str = Field(
        min_length=1,
        max_length=300,
    )

    description: str | None = None


class CaseUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    status: CaseStatus | None = None

    risk_score: float | None = Field(
        default=None,
        ge=0.0,
        le=1.0,
    )

    risk_level: RiskLevel | None = None

    outcome: CaseOutcome | None = None

    title: str | None = Field(
        default=None,
        min_length=1,
        max_length=300,
    )

    description: str | None = None


class Case(BaseModel):
    model_config = ConfigDict(extra="forbid")

    case_id: str

    investigation_id: str

    transaction_id: str | None = None

    customer_id: str | None = None

    account_id: str | None = None

    title: str

    description: str | None = None

    status: CaseStatus = CaseStatus.OPEN

    outcome: CaseOutcome | None = None

    risk_score: float | None = Field(
        default=None,
        ge=0.0,
        le=1.0,
    )

    risk_level: RiskLevel = RiskLevel.UNKNOWN

    fraud_type: str | None = None

    evidence_ids: list[str] = Field(
        default_factory=list,
    )

    findings: list[CaseFinding] = Field(
        default_factory=list,
    )

    decisions: list[CaseDecision] = Field(
        default_factory=list,
    )

    actions: list[CaseAction] = Field(
        default_factory=list,
    )

    related_cases: list[CaseMemoryReference] = Field(
        default_factory=list,
    )

    memory_ids: list[str] = Field(
        default_factory=list,
    )

    created_at: datetime

    updated_at: datetime

    resolved_at: datetime | None = None

    closed_at: datetime | None = None

File: FraudGraph-Agent/frontend/types/case.ts
export type CaseStatus =
  | "open"
  | "investigating"
  | "awaiting_evidence"
  | "awaiting_approval"
  | "actioned"
  | "escalated"
  | "resolved"
  | "closed";

export type CaseOutcome =
  | "confirmed_fraud"
  | "cleared"
  | "inconclusive"
  | "escalated";

export type DecisionType =
  | "allow"
  | "block"
  | "monitor"
  | "warn"
  | "request_evidence"
  | "escalate"
  | "create_report"
  | "close";

export interface CaseFinding {
  finding_id: string;
  title: string;
  description: string;
  evidence_ids: string[];
  confidence: number | null;
  created_at: string;
}

export interface CaseDecision {
  decision_id: string;
  decision_type: DecisionType;
  rationale: string;
  evidence_ids: string[];
  requires_approval: boolean;
  approved: boolean | null;
  approved_by: string | null;
  created_at: string;
}

export interface CaseAction {
  action_id: string;
  action_type: string;
  status: string;
  rationale: string | null;
  requires_approval: boolean;
  approval_id: string | null;
  result: Record<string, unknown>;
  created_at: string;
  completed_at: string | null;
}

export interface CaseMemoryReference {
  memory_id: string;
  case_id: string;
  similarity: number | null;
  relevance_reason: string | null;
}

export interface Case {
  case_id: string;
  investigation_id: string;
  transaction_id: string | null;
  customer_id: string | null;
  account_id: string | null;
  title: string;
  description: string | null;
  status: CaseStatus;
  outcome: CaseOutcome | null;
  risk_score: number | null;
  risk_level: string;
  fraud_type: string | null;
  evidence_ids: string[];
  findings: CaseFinding[];
  decisions: CaseDecision[];
  actions: CaseAction[];
  related_cases: CaseMemoryReference[];
  memory_ids: string[];
  created_at: string;
  updated_at: string;
  resolved_at: string | null;
  closed_at: string | null;
}

export interface CaseCreate {
  investigation_id: string;
  transaction_id?: string | null;
  customer_id?: string | null;
  account_id?: string | null;
  title: string;
  description?: string | null;
}

export interface CaseUpdate {
  status?: CaseStatus;
  risk_score?: number | null;
  risk_level?: string;
  outcome?: CaseOutcome | null;
  title?: string;
  description?: string | null;
}
